fused_moe: SM89 tactics fail on SM120 Blackwell for pure FP8 MoE (shared memory assertion)

## Environment

| Component | Version |
|-----------|---------|
| GPU | 8× NVIDIA RTX PRO 6000 Blackwell Server Edition |
| Compute Capability | SM 12.0 |
| CUDA Driver / Runtime | 13.0 / 12.9 |
| vLLM | 0.17.0 |
| flashinfer | 0.6.x (bundled with vLLM 0.17.0) |
| PyTorch | 2.10.0+cu130 |
| Python | 3.12 |
| OS | Ubuntu 24.04, x86_64 |

**Model under test:** `nvidia/nemotron-3-nano` — pure FP8, TP=8 (MoE architecture)

Note: Dense FP8 models and NVFP4 MoE models hit a separate code path and are **not** affected by this issue. This is specific to **pure FP8 MoE** on SM120/SM121.

---

## Problem

During NIM server startup, the flashinfer MoE autotuner profiles CUTLASS FP8 GroupedGEMM tactics loaded from the `fused_moe_120` JIT cache. All tactics using the `cutlass::arch::Sm89` (Ada Lovelace) dispatch path fail with a shared memory assertion error on Blackwell (CC 12.0). This affects all TP workers simultaneously and repeats across multiple tactic indices per MoE layer.

The autotuner catches the failures at profiling time and skips the failing tactics, so the server does eventually start. However, it falls back to non-Blackwell-native kernels, resulting in **suboptimal MoE throughput**.

---

## Reproduction

```bash
docker run --gpus all \
  -e VLLM_USE_FLASHINFER_MOE_FP4=1 \
  nvcr.io/nvstaging/nim-internal/nemotron-nano-3:2.0.0-rc.20260311082200 \
  --model nvidia/nemotron-3-nano --quantization fp8 --tensor-parallel-size 8
```

---

## Error Output

**Startup warning (all 8 TP workers, tactic indices 19–20+):**

```
2026-03-11 16:06:48,177 - WARNING - autotuner.py:496 - flashinfer.jit:
[Autotuner]: Skipping tactic <MoERunner object> 19, due to failure while profiling:
[TensorRT-LLM][ERROR] Assertion failed: GPU lacks the shared memory resources
to run GroupedGEMM kernel
(moe_gemm_template_dispatch.h:175)
```

**Stack trace (key frames):**

```
dispatchGemmConfig<
  __nv_fp8_e4m3,           // input
  __nv_fp8_e4m3,           // weight
  __nv_bfloat16,           // output
  cutlass::arch::Sm89,     // ← Ada Lovelace target, running on Blackwell
  EpilogueOpDefault,
  GemmShape<16, 256, 128>, // tile shape
  GemmShape<16, 64, 128>   // warp shape
>
CutlassMoeFCRunner<__nv_fp8_e4m3, ...>::gemm1(...)
GemmProfilerBackend::runProfiler(...)
```

---

## Root Cause

The bug is in `moe_gemm_template_dispatch.h:793-801`:

```cpp
} else if (sm_ >= 90) {
    // For SM120+ pure FP8 MoE (not FP8 x FP4), redirect to SM89 (Ada) FP8 kernel implementations.
    if constexpr (use_fp8 && !use_wfp4afp8) {
        if (sm_ >= 120) {
            cutlass_kernels_oss::dispatchMoeGemmToCutlass<T, WeightType, ScaleBiasType,
                                                          cutlass::arch::Sm89,  // ← Ada on Blackwell
                                                          EpilogueTag>(inputs, multi_processor_count_);
            return;
        }
    }
```

This is **intentional** — no SM120-native TMA kernels exist for pure FP8 MoE (only the FP4/NVFP4 path has SM120 TMA coverage). When the SM89 kernels are dispatched on Blackwell, `GemmGrouped::maximum_active_blocks()` returns **0** for several SM89 tile shapes (e.g. `GemmShape<16, 256, 128>`) because Blackwell's shared memory architecture does not satisfy the SM89 occupancy requirements for those tile configs. The occupancy-zero assertion fires (line 175), which the autotuner catches and skips — but the surviving tactics are sub-optimal for Blackwell.

A secondary bug existed in the occupancy-query path itself: `genericMoeGemmKernelLauncher::call` was using `compute_occupancy_for_kernel<GemmKernel>()` for occupancy queries while the actual execution path used `GemmGrouped::maximum_active_blocks()`. This mismatch allowed configs with zero execution-path occupancy to pass the pre-check and only fail at profiling time.

---

## Workaround

Use NVFP4 instead of pure FP8 for MoE models on Blackwell, which takes the SM120-native TMA path and avoids the SM89 fallback entirely:

```bash
VLLM_USE_FLASHINFER_MOE_FP4=1  # use NVFP4 MoE path (Blackwell-native)
```

Or disable flashinfer MoE entirely to fall back to a non-CUTLASS kernel:

```bash
VLLM_USE_FLASHINFER_MOE_FP4=0
```

---

## Fix

PR #3032 (superseding #2764 by @dryu-nv) adds an occupancy pre-filter in `MoERunner.get_valid_tactics()`:

1. **C++ (`flashinfer_cutlass_fused_moe_binding.cu`):** Exposes `get_tactic_occupancy(tactic_id)` FFI binding using `GemmGrouped::maximum_active_blocks()` — the same occupancy path used at execution time — to safely probe each tactic without running it. Protected by `mMutex` to prevent races with concurrent `setTactic()`/`runMoe()` calls.

2. **C++ (`moe_gemm_template_dispatch.h`):** `queryOccupancyForConfig()` implementation routes through the full dispatch chain and catches unsupported configs via try/catch, returning 0 for any config that would fail.

3. **Python (`core.py`):** `get_valid_tactics()` pre-filters the tactic list by calling `get_tactic_occupancy()` on each candidate before passing them to the autotuner. Zero-occupancy tactics are dropped silently; if the FFI call fails unexpectedly, a `logger.warning` is emitted and the tactic is included as a safe fallback.

---

## Related

- #2764 — original fix by @dryu-nv (superseded by #3032)
- #2776 — NVFP4 MoE crash on GB10/SM121 (separate but related; fixed by #2798 / CUTLASS 4.4.2 upgrade)
- #2577 — NVFP4 mm_fp4 GEMM failures on SM120 (broader SM120 NVFP4 compatibility)

--- a/csrc/fused_moe/cutlass_backend/flashinfer_cutlass_fused_moe_binding.cu
+++ b/csrc/fused_moe/cutlass_backend/flashinfer_cutlass_fused_moe_binding.cu
@@ -725,6 +725,7 @@ class FusedMoeRunner : public tvm::ffi::ModuleObj {
       // Returns 0 if the tactic is not supported on the current device (e.g., SM89 tile
       // configs with insufficient shared memory when running on SM120 Blackwell).
       return Function::FromTyped([this](int64_t tactic_id) -> int64_t {
+        std::lock_guard<std::mutex> lock(mMutex);
         if (tactic_id < 0 || tactic_id >= static_cast<int64_t>(mAllProfiles.size())) {
           return 0;
         }

--- a/csrc/fused_moe/cutlass_backend/flashinfer_cutlass_fused_moe_binding.cu
+++ b/csrc/fused_moe/cutlass_backend/flashinfer_cutlass_fused_moe_binding.cu
@@ -720,6 +720,17 @@ class FusedMoeRunner : public tvm::ffi::ModuleObj {
       return Function::FromTyped([this]() -> int64_t { return mGemm1TacticCount; });
     } else if (name == "get_gemm2_tactic_count") {
       return Function::FromTyped([this]() -> int64_t { return mGemm2TacticCount; });
+    } else if (name == "get_tactic_occupancy") {
+      // Returns the max active blocks per SM for the given tactic index.
+      // Returns 0 if the tactic is not supported on the current device (e.g., SM89 tile
+      // configs with insufficient shared memory when running on SM120 Blackwell).
+      return Function::FromTyped([this](int64_t tactic_id) -> int64_t {
+        if (tactic_id < 0 || tactic_id >= static_cast<int64_t>(mAllProfiles.size())) {
+          return 0;
+        }
+        return static_cast<int64_t>(
+            mKernelRunner->queryOccupancyForConfig(mAllProfiles[tactic_id]));
+      });
     } else if (name == "run_moe") {
       return Function::FromTyped(
           [this](TensorView output, TensorView input, TensorView token_selected_experts,

--- a/csrc/nv_internal/tensorrt_llm/kernels/cutlass_kernels/include/moe_gemm_kernels.h
+++ b/csrc/nv_internal/tensorrt_llm/kernels/cutlass_kernels/include/moe_gemm_kernels.h
@@ -315,6 +315,12 @@ class MoeGemmRunner {
 
   [[nodiscard]] int getSM() const;
 
+  // Query the occupancy (max active blocks per SM) for a given GEMM configuration.
+  // Returns 0 if the configuration is not supported on the current device.
+  // This is useful for pre-filtering tactics that would fail at execution time due to
+  // insufficient shared memory resources (e.g., SM89 tile configs run on SM120 Blackwell).
+  int queryOccupancyForConfig(cutlass_extensions::CutlassGemmConfig const& config);
+
  private:
   template <typename EpilogueTag>
   void dispatchToArch(GroupedGemmInput<T, WeightType, ScaleBiasType, OutputType> inputs,

--- a/csrc/nv_internal/tensorrt_llm/kernels/cutlass_kernels/include/moe_kernels.h
+++ b/csrc/nv_internal/tensorrt_llm/kernels/cutlass_kernels/include/moe_kernels.h
@@ -455,6 +455,9 @@ class CutlassMoeFCRunnerInterface {
   virtual void setTactic(std::optional<cutlass_extensions::CutlassGemmConfig> gemm1_config,
                          std::optional<cutlass_extensions::CutlassGemmConfig> gemm2_config) = 0;
   virtual std::vector<cutlass_extensions::CutlassGemmConfig> getTactics(MoeGemmId gemm_id) = 0;
+  // Query occupancy for a GEMM config without executing the kernel.
+  // Returns 0 if the config is incompatible with the current device.
+  virtual int queryOccupancyForConfig(cutlass_extensions::CutlassGemmConfig const& config) = 0;
 
   virtual void runMoe(void const* input_activations, void const* input_sf,
                       bool const swizzled_input_sf, int const* token_selected_experts,
@@ -633,6 +636,10 @@ class CutlassMoeFCRunner : public CutlassMoeFCRunnerInterface {
 
   std::vector<cutlass_extensions::CutlassGemmConfig> getTactics(MoeGemmId gemm_id) override {
     return moe_gemm_runner_.getConfigs(gemm_id == MoeGemmId::GEMM_2 && mayHaveFinalizeFused());
+  }
+
+  int queryOccupancyForConfig(cutlass_extensions::CutlassGemmConfig const& config) override {
+    return moe_gemm_runner_.queryOccupancyForConfig(config);
   }
 
   static std::vector<cutlass_extensions::CutlassGemmConfig> getTactics(int sm, MoeGemmId gemm_id) {

--- a/csrc/nv_internal/tensorrt_llm/kernels/cutlass_kernels/moe_gemm/moe_gemm_template_dispatch.h
+++ b/csrc/nv_internal/tensorrt_llm/kernels/cutlass_kernels/moe_gemm/moe_gemm_template_dispatch.h
@@ -166,12 +166,18 @@ struct genericMoeGemmKernelLauncher {
 
       using GemmGrouped = cutlass::gemm::device::GemmGrouped<GemmKernel>;
 
+      // Use the same runtime device check for both the occupancy query path and the
+      // execution path. This is important for SM89 (Ada) kernels running on SM120 (Blackwell):
+      // pure FP8 MoE on SM120 falls back to SM89 kernels since no native SM120 FP8 GEMM
+      // kernels are available. Some SM89 tile configs have 0 occupancy on SM120 due to
+      // hardware resource differences. Using GemmGrouped::maximum_active_blocks() ensures
+      // the occupancy query returns 0 for these configs, allowing the static heuristic to
+      // correctly skip them rather than selecting a config that will fail at execution time.
+      int occupancy = std::min(2, GemmGrouped::maximum_active_blocks());
       if (inputs.occupancy != nullptr) {
-        *inputs.occupancy =
-            tensorrt_llm::cutlass_extensions::compute_occupancy_for_kernel<GemmKernel>();
+        *inputs.occupancy = occupancy;
         return;
       }
-      int occupancy = std::min(2, GemmGrouped::maximum_active_blocks());
       TLLM_CHECK_WITH_INFO(occupancy > 0,
                            "GPU lacks the shared memory resources to run GroupedGEMM kernel");
       int const threadblock_count = sm_count_ * occupancy;
@@ -1053,6 +1059,36 @@ void MoeGemmRunner<T, WeightType, OutputType, ScaleBiasType, IsMXFPX>::moeGemm(
     GroupedGemmInput<T, WeightType, ScaleBiasType, OutputType> inputs,
     TmaWarpSpecializedGroupedGemmInput hopper_inputs) {
   runGemm<cutlass_extensions::EpilogueOpDefault>(inputs, hopper_inputs);
+}
+
+template <typename T, typename WeightType, typename OutputType, typename ScaleBiasType,
+          bool IsMXFPX>
+int MoeGemmRunner<T, WeightType, OutputType, ScaleBiasType, IsMXFPX>::queryOccupancyForConfig(
+    cutlass_extensions::CutlassGemmConfig const& config) {
+  // TMA warp-specialized configs (Hopper/Blackwell native) do not use the Ampere GroupedGEMM
+  // occupancy path, so we conservatively report them as supported (occupancy > 0).
+  if (config.is_tma_warp_specialized) {
+    return 1;
+  }
+
+  // Dispatch with inputs.occupancy set to query occupancy without running the kernel.
+  // Tensor pointers (A, B, C, etc.) are not accessed when inputs.occupancy != nullptr,
+  // so they can safely be left as null.
+  int occupancy = 0;
+  GroupedGemmInput<T, WeightType, ScaleBiasType, OutputType> inputs;
+  inputs.occupancy = &occupancy;
+  inputs.gemm_config = config;
+  // Use default epilogue for the query; epilogue type does not affect occupancy.
+  // Pass a default-constructed TmaWarpSpecializedGroupedGemmInput (isValid() == false)
+  // so that no TMA warp-specialized paths are taken for Ampere-style configs.
+  try {
+    runGemm<cutlass_extensions::EpilogueOpDefault>(inputs, TmaWarpSpecializedGroupedGemmInput{});
+  } catch (...) {
+    // If the dispatch throws (e.g., assertion failure for unsupported config/arch combination),
+    // the config is incompatible with this device — report occupancy 0.
+    return 0;
+  }
+  return occupancy;
 }
 
 }  // namespace tensorrt_llm::kernels::cutlass_kernels

--- a/flashinfer/fused_moe/core.py
+++ b/flashinfer/fused_moe/core.py
@@ -327,11 +327,7 @@ def get_cutlass_fused_moe_module(backend: str = "100", use_fast_build: bool = Fa
                 get_occ = self.fused_moe_runner.get_tactic_occupancy
             except AttributeError:
                 # get_tactic_occupancy not available in this build; skip pre-filtering
-                return all_tactics
-
-            import logging as _logging
-
-            _logger = _logging.getLogger(__name__)
+                return all_tactics if all_tactics else [-1]
 
             valid_tactics = []
             for t in all_tactics:
@@ -341,7 +337,7 @@ def get_cutlass_fused_moe_module(backend: str = "100", use_fast_build: bool = Fa
                 except Exception as e:
                     # If the query fails unexpectedly, include the tactic and let
                     # the autotuner handle any errors during profiling.
-                    _logger.warning(
+                    logger.warning(
                         "get_tactic_occupancy failed for tactic %d: %s; including in autotuner",
                         t,
                         e,

--- a/flashinfer/fused_moe/core.py
+++ b/flashinfer/fused_moe/core.py
@@ -312,10 +312,35 @@ def get_cutlass_fused_moe_module(backend: str = "100", use_fast_build: bool = Fa
 
             stage = getattr(self, "gemm_idx_for_tuning", None)
             if stage == 1:
-                return list(range(gemm1_count))
-            if stage == 2:
-                return list(range(gemm1_count, gemm1_count + gemm2_count))
-            return list(range(total))
+                all_tactics = list(range(gemm1_count))
+            elif stage == 2:
+                all_tactics = list(range(gemm1_count, gemm1_count + gemm2_count))
+            else:
+                all_tactics = list(range(total))
+
+            # Pre-filter tactics with zero occupancy on the current device.
+            # This eliminates tactics that would fail during profiling with
+            # "GPU lacks the shared memory resources" errors — notably, SM89 (Ada)
+            # tile configs used as fallback for pure FP8 MoE on SM120 (Blackwell CC 12.0)
+            # where native SM120 FP8 MoE GEMM kernels are not yet available.
+            try:
+                get_occ = self.fused_moe_runner.get_tactic_occupancy
+            except AttributeError:
+                # get_tactic_occupancy not available in this build; skip pre-filtering
+                return all_tactics
+
+            valid_tactics = []
+            for t in all_tactics:
+                try:
+                    if get_occ(t) > 0:
+                        valid_tactics.append(t)
+                except Exception:
+                    # If the query fails unexpectedly, include the tactic and let
+                    # the autotuner handle any errors during profiling.
+                    valid_tactics.append(t)
+            # Fall back to all tactics if occupancy check eliminated everything
+            # (e.g., on an unexpected architecture where all tactics report 0)
+            return valid_tactics if valid_tactics else all_tactics
 
         def forward(
             self,

--- a/flashinfer/fused_moe/core.py
+++ b/flashinfer/fused_moe/core.py
@@ -329,17 +329,30 @@ def get_cutlass_fused_moe_module(backend: str = "100", use_fast_build: bool = Fa
                 # get_tactic_occupancy not available in this build; skip pre-filtering
                 return all_tactics
 
+            import logging as _logging
+
+            _logger = _logging.getLogger(__name__)
+
             valid_tactics = []
             for t in all_tactics:
                 try:
                     if get_occ(t) > 0:
                         valid_tactics.append(t)
-                except Exception:
+                except Exception as e:
                     # If the query fails unexpectedly, include the tactic and let
                     # the autotuner handle any errors during profiling.
+                    _logger.warning(
+                        "get_tactic_occupancy failed for tactic %d: %s; including in autotuner",
+                        t,
+                        e,
+                    )
                     valid_tactics.append(t)
             # Fall back to all tactics if occupancy check eliminated everything
-            # (e.g., on an unexpected architecture where all tactics report 0)
+            # (e.g., on an unexpected architecture where all tactics report 0).
+            # If all_tactics itself is empty (zero-tactic stage), return [-1] as
+            # a sentinel so the autotuner contract is never violated with an empty list.
+            if not all_tactics:
+                return [-1]
             return valid_tactics if valid_tactics else all_tactics
 
         def forward(

